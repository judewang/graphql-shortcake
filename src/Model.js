/* eslint no-param-reassign: ["error", { "ignorePropertyModificationsFor": ["target"] }] */
import _ from 'lodash';
import fp from 'lodash/fp';
import { invokeNull } from 'thelper';
import withArrayMutator from './withArrayMutator';
import withBuilder from './withBuilder';
import withBatch from './withBatch';
import withCache from './withCache';
import withColumn from './withColumn';
import withDefaultValues from './withDefaultValues';
import withFetcher from './withFetcher';
import withGlobalId from './withGlobalId';
import withIncrementer from './withIncrementer';
import withJSONMutator from './withJSONMutator';
import withLoader from './withLoader';
import withMutator from './withMutator';
import withSearcher from './withSearcher';

class Model {
  static database = null;

  static tableName = null;

  static viewName = null;

  static idAttribute = 'id';

  static hasOperator = false;

  static hasTimestamps = true;

  static softDelete = true;

  static get(target, name) {
    if (this.has(target, name)) {
      return target.get(name);
    }

    return target[name];
  }

  static set(target, name, value) {
    if (this.has(target, name)) {
      target.set(name, value);
      return;
    }

    target[name] = value;
  }

  static get displayName() {
    return this.name;
  }

  static has(target, name) {
    return target.ownKeys.indexOf(name) > -1;
  }

  static ownKeys(target) {
    return target.ownKeys;
  }

  static getOwnPropertyDescriptor(target, name) {
    if (this.has(target, name)) {
      return { configurable: true, enumerable: true, value: target.get(name) };
    }

    return Object.getOwnPropertyDescriptor(target, name);
  }

  static refreshView = _.debounce(() => {
    this.raw(`REFRESH MATERIALIZED VIEW ${this.viewName};`);
  }, 500, { maxWait: 2000 });

  static raw(...args) {
    return this.database.raw(...args);
  }

  static forge(attributes, options) {
    const model = new this({}, options);
    return model.forge(attributes);
  }

  static fromModel(model) {
    return _.get(model, ['nativeId'], model);
  }

  static checkOperator(operator) {
    const { hasOperator } = this;
    if (!hasOperator) return null;
    if (!operator) throw new Error('operator is required');
    return this.fromModel(operator);
  }

  [Symbol.toStringTag] = 'Model';

  [Symbol.toPrimitive](hint) {
    if (hint === 'number') return Number.NaN;
    return this.toString();
  }

  constructor() {
    if (this.initialize) this.initialize();
    const handler = _.pick(this.constructor, [
      'get', 'set', 'has', 'ownKeys', 'getOwnPropertyDescriptor',
    ]);
    return new Proxy(this, handler);
  }

  initialize() {
    this._ = {};
    this.current = {};
    this.previous = {};
  }

  get isNew() {
    const { idAttribute } = this.constructor;
    return !this[idAttribute];
  }

  get values() {
    return { ...this.previous, ...this.current };
  }

  get(paths) {
    const { values } = this;
    return _.get(values, paths, null);
  }

  set(paths, value) {
    _.set(this.current, paths, value);
  }

  valueOf(name) {
    if (name) return _.get(this.current, [name], this.previous[name]);

    const { idAttribute } = this.constructor;
    const id = this.valueOf(idAttribute);
    return id ? String(id) : id;
  }

  toString() {
    const { idAttribute } = this.constructor;
    return this[idAttribute];
  }

  clone(options) {
    const model = new this.constructor({}, options);
    model.current = _.cloneDeep(this.current);
    model.previous = this.previous;
    return model;
  }

  forge(attributes) {
    return invokeNull((attrs) => {
      const values = _.get(attrs, ['previous'], attrs);
      this.current = _.cloneDeep(values);
      this.previous = values;
      return this;
    }, attributes);
  }

  merge(attributes) {
    return invokeNull((attrs) => {
      const { format } = this.constructor;
      const values = format(attrs);
      _.forEach([this.current, this.previous], data => _.assign(data, values));
      return this;
    }, attributes);
  }
}

export default fp.compose(
  withArrayMutator,
  withJSONMutator,
  withIncrementer,
  withSearcher,
  withDefaultValues,
  withCache,
  withLoader,
  withFetcher,
  withMutator,
  withBatch,
  withBuilder,
  withColumn,
  withGlobalId,
)(Model);
