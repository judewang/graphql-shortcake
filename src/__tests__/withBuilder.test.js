import _ from 'lodash';
import { client } from 'knex';
import database from './database';
import Model from '..';

class Builder extends Model {
  static database = database;

  static tableName = 'builder';

  static columns = {
    name: { type: String },
    price: { type: Number },
  }
}

class BuilderView extends Builder {
  static database = database;

  static viewName = 'builder_view';

  static columns = {
    name: { type: String },
    double: { type: Number },
  }
}

describe('Builder', () => {
  beforeAll(async () => {
    await database.raw('DROP VIEW IF EXISTS builder_view');
    await database.schema.dropTableIfExists('builder');

    await database.schema.createTable('builder', (table) => {
      table.increments();
      table.string('name');
      table.integer('price');
      table.boolean('enable');
      table.timestamps();
      table.timestamp('deleted_at');
      table.integer('created_by');
      table.integer('updated_by');
      table.integer('deleted_by');
    });

    await database.raw('CREATE VIEW builder_view AS SELECT id, price * 2 as double FROM builder');

    await new Builder().insert({ name: 'apple', price: 10 });
    await new Builder().insert({ name: 'banana', price: 20 });
  });

  it('static mutate', () => {
    expect(Builder.mutate.toString()).toBe('select * from "builder"');
    expect(BuilderView.mutate.toString()).toBe('select * from "builder"');
  });

  it('static query', () => {
    expect(Builder.query.toString()).toBe('select * from "builder"');
    expect(BuilderView.query.toString()).toBe('select * from "builder"');
  });

  it('toSQL', () => {
    const model = new Builder();
    expect(model.toSQL().sql).toBe('select id from "builder" where deleted_at IS NULL');
  });

  describe('table & view', () => {
    it('only table', async () => {
      expect(_.map(await new Builder().fetchAll(), 'price')).toEqual([10, 20]);
      expect(client).toMatchSnapshot();
    });

    it('table with view', async () => {
      expect(_.map(await new BuilderView().fetchAll(), 'double')).toEqual([20, 40]);
      expect(client).toMatchSnapshot();
    });
  });

  describe('join', () => {
    it('innerJoin', async () => {
      const model = new Builder();
      model.join('builder_view', 'builder.id', 'builder_view.id');

      const { builder } = model;
      await expect(Builder.exec(builder)).resolves.toMatchSnapshot();
      expect(builder).toMatchSnapshot();
      expect(client).toMatchSnapshot();
    });
  });

  describe('where', () => {
    it('where & orWhere', async () => {
      const model = new Builder();
      model.where('name', 'apple');
      model.orWhere('name', 'banana');

      const { builder } = model;
      await expect(Builder.exec(builder)).resolves.toMatchSnapshot();
      expect(builder).toMatchSnapshot();
      expect(client).toMatchSnapshot();
    });

    it('whereIn & orWhereIn', async () => {
      const model = new Builder();
      model.whereIn('id', [1, 2, 3]);
      model.whereNotIn('id', [4, 5, 6]);
      model.orWhereIn('price', [10, 20, 30]);
      model.orWhereNotIn('price', [40, 50, 60]);

      const { builder } = model;
      await expect(Builder.exec(builder)).resolves.toMatchSnapshot();
      expect(builder).toMatchSnapshot();
      expect(client).toMatchSnapshot();
    });

    it('whereIn with Model', async () => {
      const model = new Builder();

      const whereIn = new Builder({ price: 20 });
      model.whereIn('id', whereIn.toSQL());

      const { builder } = model;
      await expect(Builder.exec(builder)).resolves.toMatchSnapshot();
      expect(builder).toMatchSnapshot();
      expect(client).toMatchSnapshot();
    });

    it('whereNull & orWhereNull', async () => {
      const model = new Builder();
      model.whereNull('name');
      model.whereNotNull('name');
      model.orWhereNull('price');
      model.orWhereNotNull('price');

      const { builder } = model;
      await expect(Builder.exec(builder)).resolves.toMatchSnapshot();
      expect(builder).toMatchSnapshot();
      expect(client).toMatchSnapshot();
    });

    it('whereBetween & whereNotBetween', async () => {
      const model = new Builder();
      model.whereBetween('price', [10, 20]);
      model.whereNotBetween('price', [15, 16]);
      model.orWhereBetween('price', [40, 50]);
      model.orWhereNotBetween('price', [45, 46]);

      const { builder } = model;
      await expect(Builder.exec(builder)).resolves.toMatchSnapshot();
      expect(builder).toMatchSnapshot();
      expect(client).toMatchSnapshot();
    });

    it('whereRaw', async () => {
      const model = new Builder();
      model.whereRaw('enable IS TRUE');
      model.orWhereRaw('enable IS FALSE');
      model.where('price', '>', 666);
      model.where('price', '!=', 999);

      const { builder } = model;
      await expect(Builder.exec(builder)).resolves.toMatchSnapshot();
      expect(builder).toMatchSnapshot();
      expect(client).toMatchSnapshot();
    });
  });

  it('groupBy', async () => {
    const model = new Builder();
    model.groupBy('id');
    model.groupBy('name');
    const { builder } = model;
    await expect(Builder.exec(builder)).resolves.toMatchSnapshot();
    expect(client).toMatchSnapshot();
  });

  it('orderByAsc & orderByDesc', async () => {
    const model = new Builder();
    model.orderByAsc('id');
    model.orderByDesc('price', 'LAST');
    const { builder } = model;
    await expect(Builder.exec(builder)).resolves.toMatchSnapshot();
    expect(client).toMatchSnapshot();
  });

  it('limit & offset', () => {
    const model = new Builder();
    model.limit(10).offset(100);
    expect(model.limit()).toBe(10);
    expect(model.offset()).toBe(100);
  });
});
