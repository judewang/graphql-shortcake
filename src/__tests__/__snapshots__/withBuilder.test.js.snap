// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Builder groupBy 1`] = `
Array [
  Object {
    "createdAt": null,
    "createdBy": null,
    "deletedAt": null,
    "deletedBy": null,
    "enable": null,
    "id": 1,
    "name": "apple",
    "price": 10,
    "updatedAt": null,
    "updatedBy": null,
  },
  Object {
    "createdAt": null,
    "createdBy": null,
    "deletedAt": null,
    "deletedBy": null,
    "enable": null,
    "id": 2,
    "name": "banana",
    "price": 20,
    "updatedAt": null,
    "updatedBy": null,
  },
]
`;

exports[`Builder groupBy 2`] = `
Array [
  "select * from builder where deleted_at IS NULL group by id, name",
]
`;

exports[`Builder join innerJoin 1`] = `
Array [
  Object {
    "createdAt": null,
    "createdBy": null,
    "deletedAt": null,
    "deletedBy": null,
    "double": 20,
    "enable": null,
    "id": 1,
    "name": "apple",
    "price": 10,
    "updatedAt": null,
    "updatedBy": null,
  },
  Object {
    "createdAt": null,
    "createdBy": null,
    "deletedAt": null,
    "deletedBy": null,
    "double": 40,
    "enable": null,
    "id": 2,
    "name": "banana",
    "price": 20,
    "updatedAt": null,
    "updatedBy": null,
  },
]
`;

exports[`Builder join innerJoin 2`] = `
Object {
  "groupBy": Array [],
  "join": Array [
    Object {
      "direction": "inner",
      "relationalKeys": Array [
        "builder.id",
        "builder_view.id",
      ],
      "table": "builder_view",
    },
  ],
  "limit": null,
  "offset": null,
  "orderBy": Array [],
  "select": Array [
    "*",
  ],
  "skipLocked": false,
  "table": "builder",
  "view": null,
  "where": Array [],
}
`;

exports[`Builder join innerJoin 3`] = `
Array [
  "select * from builder inner join builder_view on builder.id = builder_view.id where deleted_at IS NULL",
]
`;

exports[`Builder orderByAsc & orderByDesc 1`] = `
Array [
  Object {
    "createdAt": null,
    "createdBy": null,
    "deletedAt": null,
    "deletedBy": null,
    "enable": null,
    "id": 1,
    "name": "apple",
    "price": 10,
    "updatedAt": null,
    "updatedBy": null,
  },
  Object {
    "createdAt": null,
    "createdBy": null,
    "deletedAt": null,
    "deletedBy": null,
    "enable": null,
    "id": 2,
    "name": "banana",
    "price": 20,
    "updatedAt": null,
    "updatedBy": null,
  },
]
`;

exports[`Builder orderByAsc & orderByDesc 2`] = `
Array [
  "select * from builder where deleted_at IS NULL order by id ASC, price DESC NULLS LAST",
]
`;

exports[`Builder table & view only table 1`] = `
Array [
  "select *, count(*) OVER() AS total_count from builder where deleted_at IS NULL",
]
`;

exports[`Builder table & view table with view 1`] = `
Array [
  "select *, count(*) OVER() AS total_count from builder left join builder_view on builder.id = builder_view.id where deleted_at IS NULL",
]
`;

exports[`Builder where where & orWhere 1`] = `
Array [
  Object {
    "createdAt": null,
    "createdBy": null,
    "deletedAt": null,
    "deletedBy": null,
    "enable": null,
    "id": 1,
    "name": "apple",
    "price": 10,
    "updatedAt": null,
    "updatedBy": null,
  },
  Object {
    "createdAt": null,
    "createdBy": null,
    "deletedAt": null,
    "deletedBy": null,
    "enable": null,
    "id": 2,
    "name": "banana",
    "price": 20,
    "updatedAt": null,
    "updatedBy": null,
  },
]
`;

exports[`Builder where where & orWhere 2`] = `
Object {
  "groupBy": Array [],
  "join": Array [],
  "limit": null,
  "offset": null,
  "orderBy": Array [],
  "select": Array [
    "*",
  ],
  "skipLocked": false,
  "table": "builder",
  "view": null,
  "where": Array [
    Object {
      "column": "name",
      "operator": "=",
      "or": false,
      "value": "apple",
    },
    Object {
      "column": "name",
      "operator": "=",
      "or": true,
      "value": "banana",
    },
  ],
}
`;

exports[`Builder where where & orWhere 3`] = `
Array [
  "select * from builder where (name = apple or name = banana) and deleted_at IS NULL",
]
`;

exports[`Builder where whereBetween & whereNotBetween 1`] = `
Array [
  Object {
    "createdAt": null,
    "createdBy": null,
    "deletedAt": null,
    "deletedBy": null,
    "enable": null,
    "id": 1,
    "name": "apple",
    "price": 10,
    "updatedAt": null,
    "updatedBy": null,
  },
  Object {
    "createdAt": null,
    "createdBy": null,
    "deletedAt": null,
    "deletedBy": null,
    "enable": null,
    "id": 2,
    "name": "banana",
    "price": 20,
    "updatedAt": null,
    "updatedBy": null,
  },
]
`;

exports[`Builder where whereBetween & whereNotBetween 2`] = `
Object {
  "groupBy": Array [],
  "join": Array [],
  "limit": null,
  "offset": null,
  "orderBy": Array [],
  "select": Array [
    "*",
  ],
  "skipLocked": false,
  "table": "builder",
  "view": null,
  "where": Array [
    Object {
      "column": "price",
      "operator": "BETWEEN",
      "or": false,
      "value": Array [
        10,
        20,
      ],
    },
    Object {
      "column": "price",
      "operator": "NOT BETWEEN",
      "or": false,
      "value": Array [
        15,
        16,
      ],
    },
    Object {
      "column": "price",
      "operator": "BETWEEN",
      "or": true,
      "value": Array [
        40,
        50,
      ],
    },
    Object {
      "column": "price",
      "operator": "NOT BETWEEN",
      "or": true,
      "value": Array [
        45,
        46,
      ],
    },
  ],
}
`;

exports[`Builder where whereBetween & whereNotBetween 3`] = `
Array [
  "select * from builder where (price BETWEEN 10 AND 20 and price NOT BETWEEN 15 AND 16 or price BETWEEN 40 AND 50 or price NOT BETWEEN 45 AND 46) and deleted_at IS NULL",
]
`;

exports[`Builder where whereIn & orWhereIn 1`] = `
Array [
  Object {
    "createdAt": null,
    "createdBy": null,
    "deletedAt": null,
    "deletedBy": null,
    "enable": null,
    "id": 1,
    "name": "apple",
    "price": 10,
    "updatedAt": null,
    "updatedBy": null,
  },
  Object {
    "createdAt": null,
    "createdBy": null,
    "deletedAt": null,
    "deletedBy": null,
    "enable": null,
    "id": 2,
    "name": "banana",
    "price": 20,
    "updatedAt": null,
    "updatedBy": null,
  },
]
`;

exports[`Builder where whereIn & orWhereIn 2`] = `
Object {
  "groupBy": Array [],
  "join": Array [],
  "limit": null,
  "offset": null,
  "orderBy": Array [],
  "select": Array [
    "*",
  ],
  "skipLocked": false,
  "table": "builder",
  "view": null,
  "where": Array [
    Object {
      "column": "id",
      "operator": "IN",
      "or": false,
      "value": Array [
        1,
        2,
        3,
      ],
    },
    Object {
      "column": "id",
      "operator": "NOT IN",
      "or": false,
      "value": Array [
        4,
        5,
        6,
      ],
    },
    Object {
      "column": "price",
      "operator": "IN",
      "or": true,
      "value": Array [
        10,
        20,
        30,
      ],
    },
    Object {
      "column": "price",
      "operator": "NOT IN",
      "or": true,
      "value": Array [
        40,
        50,
        60,
      ],
    },
  ],
}
`;

exports[`Builder where whereIn & orWhereIn 3`] = `
Array [
  "select * from builder where (id = ANY([1,2,3]) and id != ANY([4,5,6]) or price = ANY([10,20,30]) or price != ANY([40,50,60])) and deleted_at IS NULL",
]
`;

exports[`Builder where whereIn with Model 1`] = `
Array [
  Object {
    "createdAt": null,
    "createdBy": null,
    "deletedAt": null,
    "deletedBy": null,
    "enable": null,
    "id": 2,
    "name": "banana",
    "price": 20,
    "updatedAt": null,
    "updatedBy": null,
  },
]
`;

exports[`Builder where whereIn with Model 2`] = `
Object {
  "groupBy": Array [],
  "join": Array [],
  "limit": null,
  "offset": null,
  "orderBy": Array [],
  "select": Array [
    "*",
  ],
  "skipLocked": false,
  "table": "builder",
  "view": null,
  "where": Array [
    Object {
      "column": "id",
      "operator": "IN",
      "or": false,
      "value": Builder {
        "createdAt": null,
        "deletedAt": null,
        "id": null,
        "name": null,
        "price": 20,
        "updatedAt": null,
      },
    },
  ],
}
`;

exports[`Builder where whereIn with Model 3`] = `
Array [
  "select * from builder where (id = ANY(select id from builder where (price = 20) and deleted_at IS NULL)) and deleted_at IS NULL",
]
`;

exports[`Builder where whereNull & orWhereNull 1`] = `
Array [
  Object {
    "createdAt": null,
    "createdBy": null,
    "deletedAt": null,
    "deletedBy": null,
    "enable": null,
    "id": 1,
    "name": "apple",
    "price": 10,
    "updatedAt": null,
    "updatedBy": null,
  },
  Object {
    "createdAt": null,
    "createdBy": null,
    "deletedAt": null,
    "deletedBy": null,
    "enable": null,
    "id": 2,
    "name": "banana",
    "price": 20,
    "updatedAt": null,
    "updatedBy": null,
  },
]
`;

exports[`Builder where whereNull & orWhereNull 2`] = `
Object {
  "groupBy": Array [],
  "join": Array [],
  "limit": null,
  "offset": null,
  "orderBy": Array [],
  "select": Array [
    "*",
  ],
  "skipLocked": false,
  "table": "builder",
  "view": null,
  "where": Array [
    Object {
      "column": "name",
      "operator": "=",
      "or": false,
      "value": null,
    },
    Object {
      "column": "name",
      "operator": "!=",
      "or": false,
      "value": null,
    },
    Object {
      "column": "price",
      "operator": "=",
      "or": true,
      "value": null,
    },
    Object {
      "column": "price",
      "operator": "!=",
      "or": true,
      "value": null,
    },
  ],
}
`;

exports[`Builder where whereNull & orWhereNull 3`] = `
Array [
  "select * from builder where (name IS NULL and name IS NOT NULL or price IS NULL or price IS NOT NULL) and deleted_at IS NULL",
]
`;

exports[`Builder where whereRaw 1`] = `Array []`;

exports[`Builder where whereRaw 2`] = `
Object {
  "groupBy": Array [],
  "join": Array [],
  "limit": null,
  "offset": null,
  "orderBy": Array [],
  "select": Array [
    "*",
  ],
  "skipLocked": false,
  "table": "builder",
  "view": null,
  "where": Array [
    Object {
      "column": Array [
        "enable IS TRUE",
        undefined,
      ],
      "operator": "=",
      "or": false,
      "value": undefined,
    },
    Object {
      "column": Array [
        "enable IS FALSE",
        undefined,
      ],
      "operator": "=",
      "or": true,
      "value": undefined,
    },
    Object {
      "column": "price",
      "operator": ">",
      "or": false,
      "value": 666,
    },
    Object {
      "column": "price",
      "operator": "!=",
      "or": false,
      "value": 999,
    },
  ],
}
`;

exports[`Builder where whereRaw 3`] = `
Array [
  "select * from builder where (enable IS TRUE or enable IS FALSE and price > 666 and price != 999) and deleted_at IS NULL",
]
`;
